### DISCUSSION TITLE
How to run a script for a shortcode extension?

### DISCUSSION THREAD

**Discussion by astrowonk** [quarto-dev/quarto-cli#11630] [Q&A] [answered]
### Description

I have created my first shortcode extension! It adds [bootstrap tooltips](https://github.com/astrowonk/quarto_tooltips) over a bootstrap info icon. It works, but I wondered if it can be easier.

Bootstrap tooltips have to be initialized. Per the [bootstrap documentation](https://getbootstrap.com/docs/5.3/components/tooltips/#enable-tooltips) I can do this in quarto with some YAML on the page that wants to use them:

```yaml
include-after-body:
  text: |
    <script>
    const tooltipTriggerList = document.querySelectorAll('[data-bs-toggle="tooltip"]');
    const tooltipList = [...tooltipTriggerList].map(tooltipTriggerEl => new bootstrap.Tooltip(tooltipTriggerEl))
    </script>
```

But is there a way to do this inside the extension so users don't have to include this in their `.qmd` files? I made a javascript file with these commands in them and tried to use them via:

```{lua}
local function ensure_html_deps()
  quarto.doc.add_html_dependency({
    name = 'load_tooltips',
    version = '1.0.0',
    scripts = { "load_tooltips.js" }
  })
end
``` 

in my lua code but, that didn't work. (I did try this function in my `return {}` portion of the lua file but since it didn't work, so I removed it. I realize I now define the `ensure_html_deps` function but don't use it.)

**Comment by mcanouil** [comment]
Defining a function does not make it used.

The only code used in your Lua file is:
- https://github.com/astrowonk/quarto_tooltips/blob/main/_extensions/tooltips/tooltips.lua#L9-L18

You don't even call your `ensure_html_deps()` thus the fact that nothing happens.

It's unclear with your current code how you can tell "it does not work" as there is nothing to reproduce.

Edit: some examples

- https://github.com/mcanouil/quarto-iconify/blob/main/_extensions/iconify/iconify.lua
- https://github.com/mcanouil/quarto-elevator/blob/main/_extensions/elevator/elevator.lua

**Reply by astrowonk** [reply]
I wrote that I tried the function, it did not work, and I know left it there unused. Here is what I wrote:

>  (I did try this function in my return {} portion of the lua file but since it didn't work, so I removed it. I realize I now define the ensure_html_deps function but don't use it.)

**Comment by mcanouil** [comment]
Additionally, you are not defining/inserting your JavaScript the same way and in the same place, *i.e.*, "after body", see https://quarto.org/docs/extensions/lua-api.html#html-dependencies

**Reply by astrowonk** [reply, marked as answer]
This is the key step. The javascript needs to run after the body has been created. I must use the `afterBody` kwarg. So, changing the function to:

```
local function ensure_html_deps()
  quarto.doc.add_html_dependency({
    name = 'load_tooltips',
    version = '1.0.0',
    scripts = { { path = "load_tooltips.js", afterBody = true } }
  })
end
``` 

and then using the function inside my main `return {}` worked.
